File: php-interface/src/contract-interactor.php
<?php

namespace Uploader;

use Web3\Web3;
use Web3\Contract;
use Web3\Providers\HttpProvider;
use Web3\RequestManagers\HttpRequestManager;
use Dotenv\Dotenv;

class ContractInteractor
{
    private $web3;
    private $contract;
    private $account;

    public function __construct()
    {
        $dotenv = Dotenv::createImmutable(__DIR__ . '/../');
        $dotenv->load();

        $rpcUrl = $_ENV['RPC_URL'];
        $contractAddress = $_ENV['CONTRACT_ADDRESS'];
        $privateKey = $_ENV['PRIVATE_KEY'];

        $provider = new HttpProvider(new HttpRequestManager($rpcUrl, 10));
        $this->web3 = new Web3($provider);
        $this->account = $this->web3->eth->account->privateKeyToAccount($privateKey);

        // Load the ABI from a file or use a hardcoded ABI
        $abi = file_get_contents(__DIR__ . '/abi.json');
        $this->contract = new Contract($provider, $abi);
        $this->contract->at($contractAddress);
    }

    public function storeHashInContract(string $ipfsHash): string
    {
        $nonce = $this->web3->eth->getTransactionCount($this->account->getAddress(), 'latest');

        $transaction = $this->contract->send('storeIPFSHash', $ipfsHash, [
            'from' => $this->account->getAddress(),
            'nonce' => $nonce,
            'gas' => '2000000',
            'gasPrice' => '5000000000', // 5 Gwei
        ], $this->account->getPrivateKey());

        return $transaction->getTransactionHash();
    }
}

File: rust-interface/src/contract_interactor.rs
use std::env;
use web3::contract::{Contract, Options};
use web3::transports::Http;
use web3::types::{Address, U256};
use dotenv::dotenv;

pub async fn store_hash_in_contract(ipfs_hash: &str) -> Result<String, Box<dyn std::error::Error>> {
    dotenv().ok();
    let rpc_url = env::var("RPC_URL")?;
    let private_key = env::var("PRIVATE_KEY")?;
    let contract_address = env::var("CONTRACT_ADDRESS")?;

    let transport = Http::new(&rpc_url)?;
    let web3 = web3::Web3::new(transport);
    
    let contract = Contract::from_json(
        web3.eth(),
        contract_address.parse::<Address>()?,
        include_bytes!("abi.json"),
    )?;

    let account = web3::signing::Key::from_hex(&private_key)?;
    let nonce = web3.eth().transaction_count(account.address(), None).await?;

    let tx = contract
        .call("storeIPFSHash", (ipfs_hash.to_string(),), account.address(), Options::default())
        .await?;

    Ok(format!("{:?}", tx))
}
